File: javascript/hash/349. Intersection of Two Arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersection = function(nums1, nums2) {
    const set1 = new Set(nums1);
    const res = [];

    for (n of nums2) {
        if (set1.has(n)) {
            res.push(n);
            set1.delete(n);
        }
    }
    return res;
};

// a more elegant way
var intersection = function(nums1, nums2) {
    const set1 = new Set(nums1)
    return [...new Set(nums2)].filter((val) => set1.has(val))
};

File: python/hash/349. Intersection of Two Arrays.py
'''
own (1 set)
Time: O(m + n)
Space: O(m + n)
'''

class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        numSet = set(nums1)
        res = []

        for num in nums2:
            if num in numSet:
                res.append(num)
                numSet.remove(num)

        return res